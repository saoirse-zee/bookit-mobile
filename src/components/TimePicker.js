import React from 'react'
import { Modal, Button, DatePickerIOS, Text, View, StyleSheet } from 'react-native'
import { DateTime } from 'luxon'
import PickerButton from './PickerButton'

const getDate = date => (date ? date.toJSDate() : undefined)

export default class TimePicker extends React.Component {
  state = {
    date: this.props.date,
    isPickerVisible: false,
  }


  render() {
    const {
      label,
      onDateChange,
      minimumDate,
      maximumDate,
      bookableTimeZone,
    } = this.props
    const { date } = this.state
    const formattedStartTime = date.toLocaleString({
      hour: 'numeric',
      minute: '2-digit',
      timeZoneName: 'short',
    })
    return (
      <View>
        <PickerButton
          label={label}
          value={`${formattedStartTime}`}
          onPress={() => { this.setState({ isPickerVisible: true }) }}
        />

        <Modal visible={this.state.isPickerVisible}>
          <Text style={styles.title}>{label} time</Text>
          <DatePickerIOS
            date={this.state.date.toJSDate({ zone: bookableTimeZone })}
            timeZoneOffsetInMinutes={this.state.date.offset}
            onDateChange={(value) => {
              const newDate = DateTime.fromJSDate(value, { zone: bookableTimeZone })
              console.log(newDate.toLocaleString({
                hour: 'numeric',
                minute: '2-digit',
                timeZoneName: 'short',
              }))
              this.setState({ date: newDate })
            }}
            minuteInterval={15}
            minimumDate={getDate(minimumDate)}
            maximumDate={getDate(maximumDate)}
          />
          <Button
            title="OK"
            onPress={() => {
              this.setState({ isPickerVisible: false })
              onDateChange(this.state.date)
            }}
            color="#dddddd"
          />
        </Modal>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  title: {
    marginTop: 60,
    marginLeft: 30,
    fontSize: 30,
    fontWeight: 'bold',
    marginBottom: 50,
  },
})
