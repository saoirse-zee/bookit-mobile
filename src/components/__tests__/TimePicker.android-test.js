import React from 'react'
import moment from 'moment'
import 'moment-timezone'
import { shallow } from 'enzyme'
// eslint-disable-next-line import/no-extraneous-dependencies
import renderer from 'react-test-renderer'
// import { DatePickerAndroid, TimePickerAndroid } from 'react-native'
import TimePicker from '../TimePicker.android'

const mockDatePicker = { open: jest.fn() }
const mockTimePicker = { open: jest.fn() }
jest.mock('DatePickerAndroid', () => mockDatePicker)
jest.mock('TimePickerAndroid', () => mockTimePicker)


describe('Android Time Picker', () => {
  const date = moment('2017-12-12T22:56:54')

  beforeEach(() => {
    mockDatePicker.open.mockReturnValue({
      action: 'dateSetAction',
      year: 2018,
      month: 11,
      day: 25,
    })

    mockTimePicker.open.mockReturnValue({
      action: 'timeSetAction',
      hour: 12,
      minute: 30,
    })
  })
  it('displays the api url and nonce when these values are provided', async () => {
    const tree = renderer.create(<TimePicker
      label="Start"
      date={date}
      onDateChange={() => undefined}
      bookableTimeZone="America/New_York"
    />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('does not change the date if the user cancels date selection', async () => {
    const wrapper = shallow(<TimePicker
      label="Start"
      date={date}
      onDateChange={() => undefined}
      bookableTimeZone="America/New_York"
    />)
    mockDatePicker.open.mockReturnValue(Promise.resolve({ action: 'NOT dateSetAction' }))
    wrapper.instance().setState = jest.fn()
    await wrapper.instance().showDatePicker()
    expect(wrapper.instance().setState).not.toHaveBeenCalled()
  })

  it('does not change the date if the user cancels time selection', async () => {
    const wrapper = shallow(<TimePicker
      label="Start"
      date={date}
      onDateChange={() => undefined}
      bookableTimeZone="America/New_York"
    />)
    mockTimePicker.open.mockReturnValue(Promise.resolve({ action: 'NOT timeSetAction' }))
    wrapper.instance().setState = jest.fn()
    await wrapper.instance().showDatePicker()
    expect(wrapper.instance().setState).not.toHaveBeenCalled()
  })

  it('it updates itself with the new date when the user has selected a date', async () => {
    const wrapper = shallow(<TimePicker
      label="Start"
      date={date}
      onDateChange={() => undefined}
      bookableTimeZone="America/New_York"
    />)
    await wrapper.instance().showDatePicker()
    const expected = moment()
      .tz('America/New_York')
      .year(2018)
      .month(11)
      .date(25)
      .hour(12)
      .minute(30)
      .second(0)
      .millisecond(0)
    expect(wrapper.instance().state.date).toEqual(expected)
  })

  it('fires off the passed in "onDateChange" function after the user has picked a date', async () => {
    const onDateChanged = jest.fn()
    const wrapper = shallow(<TimePicker
      label="Start"
      date={date}
      onDateChange={onDateChanged}
      bookableTimeZone="America/New_York"
    />)
    await wrapper.instance().showDatePicker()
    const expected = moment()
      .tz('America/New_York')
      .year(2018)
      .month(11)
      .date(25)
      .hour(12)
      .minute(30)
      .second(0)
      .millisecond(0)
    expect(onDateChanged).toHaveBeenCalledWith(expected)
  })
})
