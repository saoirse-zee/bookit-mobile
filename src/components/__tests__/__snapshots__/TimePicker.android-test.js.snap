// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Android Time Picker displays the api url and nonce when these values are provided 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
      },
      Object {
        "margin": 15,
      },
    ]
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flexDirection": "row",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "width": 60,
        }
      }
    >
      Start
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#4F97FF",
          "fontSize": 18,
        }
      }
    >
      10:56 PM 
    </Text>
  </View>
</View>
`;
